Add documentation about differences in shadows across platforms

lyzhan7 opened the pull request:
Added notes and screenshots from investigation into whether the blur/2 was an iOS specific issue or a Figma specific issue, as well as some helpful resources used during the investigation.

lyzhan7 committed:
Page on how platforms handle shadows differently

lyzhan7 committed:
Add link to new Shadow page

harrieshin commented on Shadows.md:
```diff
@@ -0,0 +1,19 @@
+
+### Differences in Shadows across platforms
+- Figma uses the same shadow properties as web: i.e. the blur property
+- react-native-iOS uses shadowRadius. In order to get iOS shadows to look the same as Figma, shadowRadius should be blur/2. Native iOS should also use blur/2, as well as react-native-macOS
```
what about native macos?

lyzhan7 replied:
Did not test on native macOS or native iOS, but I can do that

Saadnajmi replied:
I would de-emphasize "React Native" per say, and maybe just say "Apple (iOS and macOS), Android, and Web". In a  separate sentence, we can say "This applies to these platforms on React Native as well (React Native, React Native macOS, and React Native Web), with a link to your snack

lyzhan7 replied:
That's a good way to put it, will do. Yeah I should add the actual snack I used, just need to clean it up first

Saadnajmi replied:
Or maybe even have a section named "What about React Native? :P

lyzhan7 replied:
Add macOS, links to the snack I used, and info about react native throughout. If you think a section dedicated to RN is still needed lmk

harrieshin commented on Shadows.md:
```diff
@@ -0,0 +1,19 @@
+
+Resources
+- can play around with web shadows [here](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Backgrounds_and_Borders/Box-shadow_generator)
+- can play around with react native iOS/Android/Web shadows [here](https://snack.expo.dev/)
```
shouldn't it be just iOS / Android? web is your previous bullet point?

lyzhan7 replied:
There's react-native on web too - I don't think it's commonly used, but it was an available option in https://snack.expo.dev/

lyzhan7 replied:
I will clarify in the documentation. Was thinking about not mentioning react-native-web altogether since it is confusing, but it was just easy to test on it since snack.expo.dev has it as an option, and have another point of comparison.

harrieshin commented:
can we add code snippets for native / react native of examples?

lyzhan7 commented:
> can we add code snippets for native / react native of examples?

Yep will do that's a good idea

Saadnajmi commented:
I think the information is good and most of my nits are about formatting. Most other pages have a "Rationale / Convention / Usage" pattern and are paragraphs rather than bullet points. Also I think we should use Apple and _Web_, rather than Sigma. I think we can word it like "Apple compared to the web platform (e.g Figma and Sketch)"

lyzhan7 commented:
> I think the information is good and most of my nits are about formatting. Most other pages have a "Rationale / Convention / Usage" pattern and are paragraphs rather than bullet points. Also I think we should use Apple and _Web_, rather than Sigma. I think we can word it like "Apple compared to the web platform (e.g Figma and Sketch)"

I noticed the Convention/Rationale/Examples/Exceptions pattern too but wasn't sure when I started if that was the best format for this information & just wanted to get what I had in my head down. Thinking about it again now, I guess the convention is that for apple platforms, blur should be divided by 2, I can stick all the screenshots into rationale/examples, usage I can link the FURN PR where we divide by 2 (and FUA if we have it). Is that what you were thinking?

Saadnajmi commented:
> > I think the information is good and most of my nits are about formatting. Most other pages have a "Rationale / Convention / Usage" pattern and are paragraphs rather than bullet points. Also I think we should use Apple and _Web_, rather than Sigma. I think we can word it like "Apple compared to the web platform (e.g Figma and Sketch)"
> 
> I noticed the Convention/Rationale/Examples/Exceptions pattern too but wasn't sure when I started if that was the best format for this information & just wanted to get what I had in my head down. Thinking about it again now, I guess the convention is that for apple platforms, blur should be divided by 2, I can stick all the screenshots into rationale/examples, usage I can link the FURN PR where we divide by 2 (and FUA if we have it). Is that what you were thinking?

I wouldn't link PRs, but instead have minimal code examples. My current idea of an outline is this, feel free to differ:

```
Convention:
"If you are trying to align with designs made with a web tool (like Figma or Sketch), divide by 2.  This applies to React Native as well.

Rationale:
 See these screenshots. It's not clear why, but shadows are rendered differently on iOS vs Web vs Android. 
 
Examples:
 Probably React Native Code / Snack here?
 Probably also the FluentUI Apple helper function here
 
```

lyzhan7 committed:
Address some comments

- Made a bigger table, added code + links
- Added some context
- Changed sections

Saadnajmi commented on Shadows.md:
```diff
@@ -0,0 +1,27 @@
+
+## Rationale
+
+Here is a comparison of shadow blur/radius across different platforms:
```
Maybe something more like 

"We compared how shadows are rendered across Platforms, and found a discrepancy in how blur is handled. For whatever reason, divine by 2 seems to have the correct effect. See the table below:

And you can either keep the table in this section, or move it to examples. I'm not too strongly opinionated either way.

lyzhan7 committed:
Add example + additional work

Saadnajmi commented on Shadows.md:
````diff
@@ -0,0 +1,62 @@
+
+## Examples
+
+In Fluent UI React Native, here is a snippet of the code that takes a shadowToken object (basically an object that describes shadows the way web does, with blur) and converts it into a react native shadow.
+
+```typescript
+const appleShadowBlurAdjustment = 0.5;
+const defaultShadowBlurAdjustment = 1;
+
+function getShadowTokenStyleSet(shadowToken: ShadowToken) {
+  const keyShadow = shadowToken.key;
+  const ambientShadow = shadowToken.ambient;
+  const shadowBlurAdjustment = Platform.OS === 'macos' || Platform.OS === 'ios' ? appleShadowBlurAdjustment : defaultShadowBlurAdjustment;
+
+  return {
+    key: {
+      shadowColor: shadowColorFromRGBAColor(keyShadow.color),
+      shadowOpacity: shadowOpacityFromRGBAColor(keyShadow.color),
+      shadowRadius: keyShadow.blur * shadowBlurAdjustment,
+      shadowOffset: {
+        width: keyShadow.x,
+        height: keyShadow.y,
+      },
+    },
+    ambient: {
+      shadowColor: shadowColorFromRGBAColor(ambientShadow.color),
+      shadowOpacity: shadowOpacityFromRGBAColor(ambientShadow.color),
+      shadowRadius: ambientShadow.blur * shadowBlurAdjustment,
+      shadowOffset: {
+        width: ambientShadow.x,
+        height: ambientShadow.y,
+      },
+    },
+  };
+}
+```
+
+Additional work:
+- todo: add FluentUI Apple helper function here
+- todo: add Win32 code + screenshot as well
````
I think we don't really need these sections or TODOs anymore. I don't really want to add our whole very-specific-to-fluent methods here in docs that should be Fluent agnostic. We _can_ link the actual files in FURN and FUA as examples (without copying over the whole methods), but I think the screenshot above are sufficient. 

Excellent work!

lyzhan7 replied:
Yeah I'm not really tied to adding the code, I thought I could copy over the method in FURN without it being so long but it ended up just being most of the method. I'm actually thinking I'll just get rid of it altogether, just remembered I think we're maybe going to change the token values instead of adding a helper in FURN/FUA? And the other todo I had was win32 which I will add, so I can get rid of that section altogether

Saadnajmi commented on Shadows.md:
```diff
@@ -0,0 +1,62 @@
+
+| Platform + Notes | Code | Screenshot |
+| - | - | - |
+| Web: CSS </br></br> [box-shadow documentation](https://www.w3schools.com/cssref/css3_pr_box-shadow.asp), also can try out shadows [here](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Backgrounds_and_Borders/Box-shadow_generator) | <img width="2000" alt="Screen Shot 2022-09-07 at 1 11 53 PM" src="https://user-images.githubusercontent.com/78454019/188968215-f5c612c3-970d-4ec5-8ec6-0bc5d563 ...
+| Web: Figma </br></br> Figma uses the same shadow blur property as web | <img width="500" alt="Screen Shot 2022-09-07 at 12 31 38 PM" src="https://user-images.githubusercontent.com/78454019/188961635-0411b0f6-03e2-42e4-8b27-10d4674470a6.png"> | <img width="1200" alt="Screen Shot 2022-09-06 at 1 48 25 PM" src="https://user-images.githubusercontent.com/78454019/188753714-0ddf59b5-515b-4d78-bba8-65 ...
+| Web: React Native Web </br></br> React Native shadowRadius documentation [here](https://reactnative.dev/docs/shadow-props), CSS shadow blur documentation [here](https://www.w3schools.com/cssref/css3_pr_box-shadow.asp), also can test shadows with this [snack](https://snack.expo.dev/@lyzhan/shadow-ios-web-android). Uses shadowRadius, which ends up being the same as the blur property (don't need t ...
+| Apple: React Native iOS </br></br> React Native shadowRadius documentation [here](https://reactnative.dev/docs/shadow-props), native Apple shadowRadius documentation [here](https://developer.apple.com/documentation/quartzcore/calayer/1410819-shadowradius), also can test shadows with this [snack](https://snack.expo.dev/@lyzhan/shadow-ios-web-android). In order to get Apple shadows to look the sa ...
+| Apple: Native macOS </br></br> [shadowBlurRadius](https://developer.apple.com/documentation/uikit/nsshadow/1429846-shadowblurradius) documentation | <img width="443" alt="NativeMacOS_100" src="https://user-images.githubusercontent.com/78454019/189423358-4223d549-e589-4aa3-abc7-7b1e9a6ef04e.png"> <img width="433" alt="NativeMacOS_50" src="https://user-images.githubusercontent.com/78454019/189423 ...
+| Android: React Native Android </br></br> Can test shadows with this [snack](https://snack.expo.dev/@lyzhan/shadow-ios-web-android). Uses elevation, which is just different from both how web/iOS handle shadows| <img width="311" alt="Screen Shot 2022-09-07 at 1 42 18 PM" src="https://user-images.githubusercontent.com/78454019/188983536-e4a17f29-456f-411d-968d-43e32a93f1b6.png"> | <img width="200" ...
```
These images _could_ be local images in the repo to simplify the paths (see the keyboard focus page for an example). But if this works, I guess that's alright.

lyzhan7 replied:
Yeah thinking I'll leave it for now...if it becomes an issue in the future I can change it, have all the screenshots saved locally

lyzhan7 committed:
Add win32, remove examples/todo section

lyzhan7 committed:
Remove win32 screenshots
